### DISCUSSION TITLE
About model property types, events and behaviors

### DISCUSSION THREAD

**Discussion by raicabogdan** [phalcon/cphalcon#16538] [Q&A]
Hello,

I've started an app following documentation and what was recommended. I've started making a model User with the id, first_name, last_name , created_at and updated_at. Very simple.

The properties are all typed hinted, using int for id, string for names and DateTimeImmutable for the created_at and updated_at.

They also have getters and setters , each with their appropriate type declaration, eg:

```
    protected $created_at;

    public function getCreatedAt(): ?DateTimeImmutable
    {
        return $this->created_at;
    }

    public function setCreatedAt(?DateTimeImmutable $created_at): static
    {
        $this->created_at = $created_at;

        return $this;
    }
```

I noticed that you can not make property types when you declare the properties. My initial purpose was to make them all nullable, eg:
```
    protected ?DateTimeImmutable $created_at = null;
```
Then I read that phalcon returns all fields in string format only and new Model objects are assign null by default, so I just used protected $property only. Lastly to get the conversion to the specific type, I used model events
```
    public function afterFetch(): void
    {
        $this->created_at  = DateTimeImmutable::parse($this->created_at );
    }
```
and so far, these act as I expected if the correct approach is to use property without types, setters/getters with types.

I later saw about behaviours and I can say the idea is fabulous and very easy to reuse code, yet it seems I stumble on a particular race condition thing that I'm unsure it is by design or not. Using events, I can type hint the getters and setters just fine, but if I use behaviour class, I have to set my getters as string, only then will the behaviour take charge and update the model as I expected. So when using the behaviour, my model looks like that this

```
    protected $created_at;

    public function getCreatedAt(): ?string
    {
        return $this->created_at;
    }

    public function setCreatedAt(?DateTimeImmutable $created_at): static
    {
        $this->created_at = $created_at;

        return $this;
    }
```
Otherwise you get the error:
```
Uncaught TypeError: App\Models\User::getUpdatedAt(): Return value must be of type ?\DateTimeImmutable, string returned
```
So with all the ramblin out of the way, I believe there are two ways to achieve this, while in all cases, the property is set without type, or at least set to ?string maybe. The null value is not required.

1. Using Model Behaviors, I'm unsure how it retrieves the field value since I think it should not be using the getters yet in the events.
2. Using Model Events seems to be working properly since you basically access the property with `$this->property` so it goes around the getter type.

It seems to me that I'm overcomplicating things with getters and setters, what's the proper standard way to achieve types with model, using behaviors? It seems this is the more elegant way, yet it I'm not sure what I'm doing wrong. Why do getters need to be set as ?string while setters work fine. In my Behaviour class, I do access the property as in within normal events with $model->{$field}

```
    public function notify(string $type, ModelInterface $model): void
    {
         $field = $this->options['field'];
         $model->{$field} = DateTimeImmutable::createFromFormat($model->{$field});
     }
```

Thank you, and look forward to someone explaining if there's something I did wrong.

**Comment by shelton7863** [comment]
Ohhh! i will try to make it solve. [Catlikesbest](https://www.catlikesbest.com/)

**Comment by s-ohnishi** [comment]
Aren't the only things you can exchange with a database basically numbers or strings?
For example, even if you specify `DATETIME` on the database side, what you can get and set is a `string`, and I don't think it will directly return any class.
So I think it makes sense to exchange `string` between the model and the database, and `DateTimeImmutable` between the program and the model.
I think it's the job of setters and getters to convert between strings and classes.
If you use `$this->property`, a smart IDE will probably display a warning before execution (assuming you've specified the type properly).
